staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -485,25 +500,28 @@ func (qs *queueSet) shuffleShardLocked(hashValue uint64, descr1, descr2 interfac
 	offset := qs.enqueues % handSize
 	qs.enqueues++
 	bestQueueIdx := -1
-	bestQueueSeatsSum := int(math.MaxInt32)
+	minQueueWork := math.Inf(1)
 	for i := 0; i < handSize; i++ {
 		queueIdx := hand[(offset+i)%handSize]
 		queue := qs.queues[queueIdx]
-		waitingSeats := queue.requests.SeatsSum()
-		// TODO: Consider taking into account `additional latency` of requests
-		// in addition to their seats.
-		// Ideally, this should be based on projected completion time in the
+		queueWorkEstimate := queue.requests.QueueWorkEstimate()
+
+		// the best queue is chosen based on projected completion time in the
 		// virtual world of the youngest request in the queue.
-		thisSeatsSum := waitingSeats + queue.seatsInUse
-		klog.V(7).Infof("QS(%s): For request %#+v %#+v considering queue %d of %d seats waiting and %d executing, virtualStart=%vss", qs.qCfg.Name, descr1, descr2, queueIdx, waitingSeats, queue.seatsInUse, queue.virtualStart)
-		if thisSeatsSum < bestQueueSeatsSum {
-			bestQueueIdx, bestQueueSeatsSum = queueIdx, thisSeatsSum
+		thisSeatsSum := queueWorkEstimate.RequestedSeatsSum + queue.seatsInUse
+
+		// this is the total amount of work either in progress or waiting
+		// in this queue, we will select the queue with the minimum.
+		thisQueueWork := float64(thisSeatsSum)*qs.estimatedServiceSeconds + float64(queueWorkEstimate.AdditionalWorkSum)
```

Comment:
I think the score for a queue should be the sum, over all the executing and waiting requests, of the estimated work in that request (`req.WorkEstimate.InitialSeats * qs.estimatedServiceSeconds + req.WorkEstimate.FinalSeats * req.WorkEstimate.AdditionalLatency.Seconds()`).

The current draft of this PR make a different calculation.  It uses the sum over executing and waiting requests of `req.MaxSeats() * qs.estimatedServiceSeconds` plus the sum over waiting requests of `req.WorkEstimate.FinalSeats * req.WorkEstimate.AdditionalLatency.Seconds()`.